amr-wind/helics.cpp

```diff
@@ -0,0 +1,239 @@
+    // receive turbine yaw directions (num_turbines)
+    if (amrex::ParallelDescriptor::IOProcessor()) {
+        std::stringstream charFromControlCenter;
+        currenttime = m_vfed->requestNextStep();
+        int subCount = m_vfed->getInputCount();
+        helicscpp::Input sub;
+
+        for (int i = 0; i < subCount; i++) {
+            int yy = i;
+            sub = m_vfed->getSubscription(yy);
+            charFromControlCenter << sub.getString().c_str();
+            std::cout << "\n advancing time " << charFromControlCenter.str();
+        }
+
+        if (currenttime > 1) {
+            // Igonre timestep 0 since message pipe has junk.
+            // // Unpack the values from the control center using a string
+            // stream
+
+            std::list<double> return_list;
+            const std::string comma = ",";
+            tokenize(charFromControlCenter.str(), comma, return_list);
+
+            for (int i = m_turbine_yaw_to_amrwind.size() - 1; i >= 0; --i) {
+                m_turbine_yaw_to_amrwind[i] = return_list.front();
+                return_list.pop_front();
+            }
+
+            m_inflow_wind_direction_to_amrwind = return_list.front();
+            return_list.pop_front();
+            m_inflow_wind_speed_to_amrwind = return_list.front();
+            return_list.pop_front();
+            float time = return_list.front();
+            return_list.pop_front();
+
+            std::cout << "\n speed: " << m_inflow_wind_speed_to_amrwind
+                      << "  direction: " << m_inflow_wind_direction_to_amrwind
+                      << " Time " << time << std::endl;
+            for (int i = 0; i < m_turbine_yaw_to_amrwind.size(); ++i) {
+                std::cout << "T" << i << " yaw: " << m_turbine_yaw_to_amrwind[i]
+                          << ' ';
+            }
+            std::cout << std::endl;
+        }
+
+        std::stringstream ssToControlCenter;
+
+        helicscpp::Publication pub;
+
+        int pubCount = m_vfed->getPublicationCount();
+
+        amrex::Real wind_speed = 123.0;
+        amrex::Real wind_direction = 456.0;
+
+        auto& phy_mgr = m_sim.physics_manager();
+        if (phy_mgr.contains("ABL")) {
+            auto& abl = phy_mgr.get<amr_wind::ABL>();
+            const amrex::Real height = 90.0;
+            wind_speed = abl.abl_statistics()
+                             .vel_profile()
+                             .line_hvelmag_average_interpolated(height);
+            amrex::Real velx =
+                abl.abl_statistics().vel_profile().line_average_interpolated(
+                    height, 0);
+            amrex::Real vely =
+                abl.abl_statistics().vel_profile().line_average_interpolated(
+                    height, 1);
+            const amrex::Real turbine_angle = std::atan2(vely, velx);
+            wind_direction = -amr_wind::utils::degrees(turbine_angle) + 270.0;
+        }
+
+        std::cout << "pub count: " << pubCount << std::endl;
```

Comment:
Even though we're in an `if` statement for the print, I still think we might want to use `amrex::Print()` in case people copy and paste. I'm not sure we want `std::cout` anywhere in the code.